Firestore/core/src/api/firestore.cc

```diff
@@ -242,6 +251,68 @@ DatabaseInfo Firestore::MakeDatabaseInfo() const {
                       settings_.ssl_enabled());
 }
 
+void Firestore::SetIndexConfiguration(std::string config,
+                                      util::StatusCallback callback) {
+  EnsureClientConfigured();
+
+  bundle::JsonReader reader;
+  if (!settings_.persistence_enabled()) {
+    callback(Status(Error::kErrorFailedPrecondition,
+                    "Cannot enable indexes when persistence is disabled."));
+    return;
+  }
+
+  auto json_object =
+      nlohmann::json::parse(config.begin(), config.end(),
+                            /*callback=*/nullptr, /*allow_exceptions=*/false);
+  if (json_object.is_discarded()) {
+    callback(Status(Error::kErrorDataLoss, "Invalid Json format."));
+    return;
+  }
+
+  std::vector<FieldIndex> parsed_indexes;
+  const auto& json_indexes = reader.RequiredArray("indexes", json_object);
+  for (const auto& json_index : json_indexes) {
+    const std::string& collection_group =
+        reader.RequiredString("collectionGroup", json_index);
+    std::vector<Segment> segments;
+    std::vector<json> default_vector;
+    const auto& json_fields =
+        reader.OptionalArray("fields", json_index, default_vector);
+    for (const auto& json_field : json_fields) {
+      FieldPath field_path = FieldPath::FromServerFormat(
+                                 reader.RequiredString("fieldPath", json_field))
+                                 .ValueOrDie();
+      std::string default_string;
+      if ("CONTAINS" ==
+          reader.OptionalString("arrayConfig", json_field, default_string)) {
+        segments.emplace_back(
+            Segment(std::move(field_path), Segment::Kind::kContains));
+      } else if ("ASCENDING" ==
+                 reader.OptionalString("order", json_field, default_string)) {
+        segments.emplace_back(
+            Segment(std::move(field_path), Segment::Kind::kAscending));
+      } else {
+        segments.emplace_back(
+            Segment(std::move(field_path), Segment::Kind::kDescending));
+      }
+    }
```

Comment:
I wonder if we should add a check like line 305 here so we return immediately instead of moving to the next index.